L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,450 @@
+L46: C-core: New TLS Credentials API
+----
+* Author(s): ZhenLian
+* Approver: markdroth, yashykt, jiangtaoli2016
+* Status: Draft
+* Implemented in:
+* Last updated: September 9, 2020
+
+## Abstract
+
+This proposal aims to provide the following security features in a new TLS credential API:
+1. credential reloading: reload transport credentials periodically from various sources, e.g. from credential files, or users' customized Provider implementations, without shutting down
+2. custom authorization check: optionally bypass the certificate check, bypass the hostname check on the client side, and perform an additional per-connection authorization callback at the end of authentication
```

Comment:
How about re-phrasing this paragraph as follows?

Support custom verification on server identity: 

- Perform both certificate and hostname verification checks (default option).
- Perform certificate verification but skip hostname verification check.
- Skip both certificate and hostname verification checks.

In the last two cases, custom server authorization check MUST be performed.

Reply:
SG! That's a better format of wording.
I reworded it a bit because we are now also doing custom authorization on server side, for reasons to support CRL in the future, etc. And that's the reason I replaced all the "server authorization" with "authorization" throughout this proposal.

Reply:
Oh I see. Thanks for the clarification. For the purpose of CRL check, does it matter if the check is done after TLS handshake finishes? I expect it should be done before the TLS handshake starts, but the current server authorization check is designed to be performed after the handshake. @jiangtaoli2016

Reply:
The standard revocation check is during TLS handshake (not before handshake :-) However, given gRPC core does not support CRL check yet. This custom authorization check provides an opportunity to check revocation before gRPC sends actual payload.

Reply:
Understood. My concern is that as it deviates from the standard practice (i.e., perform check after the handshake finishes), we need to be more careful about its implication before introducing this feature because once the feature is there, nothing can prevent OSS users from using it, and it is also more difficult to retire it once we have a standard CRL check.

Reply:
There are feature requests to support custom authz check on server side. see https://github.com/grpc/grpc/issues/23498. Today, on client side, we expose custom authorization check callback for the purpose of server authorization, but many customers use it for other legitimate purpose. E.g., Microsoft uses it for validating cert chain.

By exposing a method to allow application to validate certificate (e.g., certificate check, revocation check) adds significant security value. Revocation check during handshake vs revocation check after handshake has little security implication -- as far as revocation is before actual RPC payload.

Reply:
Thanks for the clarification. I misunderstood the way CRL check will be realized. I thought the intention was for gRPC to implement CRL checks (after TLS handshake finishes) and expose an API for users to use, but instead we want to introduce a similar authorization check mechanism at the server side and it's up to users to implement the callback that may do the CRL check (if they want to). If it's the latter, I am fine with it.